drivers: dac: silabs_vdac implementation

silabs-BastienB opened the pull request:
The goal of this PR is to enable the use of the DAC driver for series 2 devices.
This PR does the following:
- Fix the clock branch bindings.
- Import sl_hal_vdac in hal_silabs/simplicity_sdk.
- Enables the hal_silabs VDAC module to be compiled for series 2.
- Defines a compatible binding which can be used for series 2 parts that have a VDAC peripheral.
- Instantiates the xg2x parts with a VDAC peripheral with a VDAC node in their respective part dtsi.
- Implements the DAC driver compatible with silabs,vdac.
- Adds support for the DAC driver in the relevant series 2 development kits and radio boards.
- Extends the dac_api and dac_loopback tests for xg23 and xg24 boards.

silabs-BastienB committed:
include: zephyr: dt-bindings: clock: silabs: fix clock branch

Updated CLOCK_BRANCH bindings to stay in sync with the enum
sl_clock_branch_t in the Silicon Labs HAL.

Signed-off-by: Bastien Beauchamp <bastien.beauchamp@silabs.com>

silabs-BastienB committed:
modules: hal_silabs: enable VDAC module in series 2 hal

This enables Silabs Series 2 HAL source files for the Silabs
VDAC module to be compiled when the SOC_SILABS_ACMP
configuration is selected.

Signed-off-by: Bastien Beauchamp <bastien.beauchamp@silabs.com>

silabs-BastienB committed:
dts: bindings: dac: define silabs,vdac bindings

Defines bindings that are compatible with Silabs VDAC.
Reference your part's design book when configuring
values for the properties.

Signed-off-by: Bastien Beauchamp <bastien.beauchamp@silabs.com>

silabs-BastienB committed:
dts: arm: silabs: instantiate VDAC nodes for xg2x parts

Defines a VDAC node for xg23 and xg24 parts, which are all
compatible with the silabs,vdac binding.

Signed-off-by: Bastien Beauchamp <bastien.beauchamp@silabs.com>

silabs-BastienB committed:
drivers: dac: implement the silabs_vdac compatible driver

This implements the DAC driver for silabs VDAC peripherals
using the silabs,vdac compatible binding.

Signed-off-by: Bastien Beauchamp <bastien.beauchamp@silabs.com>

silabs-BastienB committed:
boards: silabs: update board support for DAC driver

The xg24_dk2601b, xg24_ek2703a, xg23_rb4210a and
xg24_rb4187c board yaml files were updated
to support the DAC driver.

Signed-off-by: Bastien Beauchamp <bastien.beauchamp@silabs.com>

silabs-BastienB committed:
tests: drivers: dac: add xg23 and xg24 boards to dac tests

Add board overlays and defines to the test sources files to execute
the dac_api and dac_loopback tests on xg23_rb4210a, xg24_dk2601b,
xg24_ek2703a and xg24_rb4187c.

Signed-off-by: Bastien Beauchamp <bastien.beauchamp@silabs.com>

asmellby commented on tests/drivers/dac/dac_api/src/test_dac.c:
```diff
@@ -110,7 +110,14 @@
 #elif defined(CONFIG_BOARD_MIMXRT1170_EVK)
 
-#define DAC_DEVICE_NODE		DT_NODELABEL(dac)
+#define DAC_DEVICE_NODE		DT_NODELABEL(dac0)
```
Unrelated change?

silabs-BastienB replied:
A change made by mistake. I reverted it.

asmellby commented on tests/drivers/dac/dac_api/src/test_dac.c:
```diff
@@ -110,7 +110,14 @@
+#define DAC_RESOLUTION	12
+#define DAC_CHANNEL_ID	0
+
+#elif defined(CONFIG_BOARD_XG24_DK2601B) || \
```
Can this be made more generic? E.g. `CONFIG_SOC_FAMILY_SILABS_S2`?

silabs-BastienB replied:
I made the change.

asmellby commented on tests/drivers/dac/dac_loopback/src/test_dac.c:
```diff
@@ -184,6 +184,23 @@
+#define ADC_REFERENCE		ADC_REF_VDD_1
+#define ADC_ACQUISITION_TIME	ADC_ACQ_TIME_DEFAULT
+#define ADC_CHANNEL_ID		0
+#define ADC_1ST_CHANNEL_INPUT	0x20	/* IADC_INPUT_DAC0 */
```
Include the dt-bindings header for ADC inside the silabs-specific section to avoid hard-coding the value (see the `nrf_saadc.h` inclusion above for inspiration).

asmellby reviewed (commented):
Thanks for this! I've got a few comments, nothing major.

asmellby commented on dts/arm/silabs/xg23/xg23.dtsi:
```diff
@@ -574,6 +574,15 @@
+		vdac0: vdac@59024000 {
+			compatible = "silabs,vdac";
+			reg = <0x59024000 0x4000>;
+			interrupts = 55 2>;
```
typo here causes the build error.

silabs-BastienB replied:
I fixed it.

silabs-BastienB committed:
include: zephyr: dt-bindings: dac: silabs: generate VDAC

Generates bindings that define VDAC auxiliary output.
These bindings can be used in devicetree sources to configure
the aux-output property.

Signed-off-by: Bastien Beauchamp <bastien.beauchamp@silabs.com>

silabs-BastienB commented on modules/hal_silabs/simplicity_sdk/CMakeLists.txt:
```diff
@@ -264,6 +264,8 @@ endif()
 zephyr_library_sources_ifdef(CONFIG_SILABS_SISDK_LETIMER   ${PERIPHERAL_DIR}/src/sl_hal_letimer.c)
 zephyr_library_sources_ifdef(CONFIG_SILABS_SISDK_TIMER     ${PERIPHERAL_DIR}/src/sl_hal_timer.c)
 
+zephyr_library_sources_ifdef(CONFIG_SOC_SILABS_VDAC        ${PERIPHERAL_DIR}/src/sl_hal_vdac.c)
```
I fixed this.

silabs-BastienB commented on dts/bindings/dac/silabs,vdac.yaml:
```diff
@@ -0,0 +1,169 @@
+        Set to "None" if the auxiliary outputs are used.
+      type: string
+      enum:
+        - "None"
```
I applied your suggestion. I removed `aux-output-port` and added dt-bindings for `aux-output`.

silabs-BastienB commented on drivers/dac/Kconfig.silabs_vdac:
```diff
@@ -0,0 +1,12 @@
+	default y
+	depends on DT_HAS_SILABS_VDAC_ENABLED
+	select PINCTRL
+	select SOC_SILABS_VDAC
```
I fixed this.

silabs-BastienB commented on dts/bindings/dac/silabs,vdac.yaml:
```diff
@@ -0,0 +1,169 @@
+# Copyright (c) 2025 Silicon Laboratories Inc.
+# SPDX-License-Identifier: Apache-2.0
+description: |
```
It's done.

Martinhoff-maker reviewed (commented):
Only minor comments, otherwise LGTM.
Don't forget to add a commit where you modify west.yml to point the corresponding hal_silabs pull request.
edit: also don't forget to take a look at sonarqube issue.

silabs-BastienB commented on dts/arm/silabs/xg24/xg24.dtsi:
```diff
@@ -552,6 +552,28 @@
 			status = "disabled";
 		};
+
+		vdac0: vdac@59024000 {
```
I fixed it. I also updated west.yml to point my hal_silabs pull request.
I checked the sonarqube issues, but I cannot find any fix. This same issue is present 25 times in Zephyr.

martinjaeger commented on include/zephyr/dt-bindings/dac/silabs-vdac.h:
Wrong commit message title (replace comparator with dac).

silabs-BastienB replied:
I fixed it. Thank you for the review.

martinjaeger commented on tests/drivers/dac/dac_api/boards/xg23_rb4210a.overlay:
Instead of adding the overlays for individual boards here, can you please add the relevant devicetree nodes directly to the board's .dts?

martinjaeger replied:
We are choosing several other sensible defaults in board files already (e.g. clock prescalers, UART baud rates, etc.), so I don't think it's an issue to set a default voltage reference. The user can easily override it in an overlay for a custom application.

martinjaeger replied:
Our current approach (at least for the DAC) has been to provide a useful configuration in Devicetree on board level with `status = "okay"`, but keep it disabled in Kconfig. That's also my understanding of the "left disabled" in above quote from the docs.

This approach allows the user to easily identify the name of the DAC node and a suitable base configuration by looking at the DTS file, without searching through tests, SOC DTS or bindings. In addition to that, twister can automatically build a test based on the "supported" flag in the `board-name.yaml` file.

If we were to add more tests for the DAC, we would have to copy the overlay files to each of them, which doesn't scale very well. I realize that we still need to provide an overlay for the loopback test, though, as this test requires both ADC and DAC to be enabled.

As far as I can see, a user would not need to `/delete-property/` for the default DAC configuration. He/she would only need to override `voltage-reference` and/or `aux-output` if desired.

martinjaeger commented on drivers/dac/dac_silabs_vdac.c:
```diff
@@ -0,0 +1,155 @@
+	}
+
+	/* Write value to VDAC channel */
+	sl_hal_vdac_set_output_channel(config->base, channel, value);
```
Value should be checked if it's within the allowed range.

martinjaeger commented on dts/arm/silabs/xg23/xg23.dtsi:
```diff
@@ -582,6 +582,24 @@
+			reg = <0x59024000 0x4000>;
+			interrupts = <55 2>;
+			clocks = <&cmu CLOCK_VDAC0 CLOCK_BRANCH_VDAC0CLK>;
+			status = "okay";
```
It should be kept `"disabled"` here and enabled (`status = "okay"`) in the board-level DTS. Please see also how it's done for other DAC drivers, e.g. STM32.

silabs-BastienB replied:
Alright, it should be fixed now.

silabs-BastienB committed:
boards: silabs: enable VDAC nodes for xg2x boards

Set status to okay for VDAC nodes for xg23 and xg24 boards.

Signed-off-by: Bastien Beauchamp <bastien.beauchamp@silabs.com>

jerome-pouiller reviewed (dismissed):
You need to rebase https://github.com/zephyrproject-rtos/hal_silabs/pull/121 then relaunch the CI.

Otherwise, LGTM.

jhedberg commented:
The `west.yml` change will need to be coordinated with #92808, since that one bumps `hal_silabs` to latest `main` whereas this bumps it to `main~2`
